---
title: "ORFik Overview"
author: "Haakon Tjeldnes"
date: "`r BiocStyle::doc_date()`"
package: "`r pkg_ver('ORFik')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{ORFik Overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Welcome to the ORFik package. This vignette will walk you through our main package usage with examples.
ORFik is an R package containing various functions for analysis of Ribo-Seq and RNA-Seq data.
ORFik currently supports:

- Finding Open Reading Frames (very fast) in the genome of interest or on the set of transcripts.
- Metaplots for Ribo-Seq allowing to spot the shift.
- Shifting functions for the Ribo-Seq data.
- Various measurements of gene identity e.g. FLOSS, coverage, ORFscore, entropy.
- Utility functions to extend GenomicRanges, e.g. the functions tile1 and groupGRangesBy.


# Introduction

`ORFik` main usage is to get Open reading frames compatitable with Bioconductor's GenomicRanges package, combined with Ribo-seq, RNA-seq and cage-seq data.
You can then get the features you want, and make various plots. 

# Finding ORFs:
To find all Open Reading Frames (ORFs) use ?findORFs function. If input is only .fasta file (or BSGenome), findORFs will treat it as prokaryotic genome and findORFs on the chromosomes assuming no splicing. If .gtf file is also supplied, sequences of transcripts from .gtf file will be extracted and findORFs will search there. 
It is also possible to operate on a vector of characters or Biostrings::DNAString or Biostrings::DNAStringSet.

We will now provide two examples of finding ORFs, one with character vector and one with DNAStringSet from BSGenome.

# Character vector example

```r
library(ORFik)

# Make Character vector example data, here we dont use fasta file, but just a character vector:
seqname <- c("tx1","tx1") # <- a two exon transcript
seqs <- c("ATGATGTAATAA") # <- simple sequence
grIn1 <- GRanges(seqnames = rep("1", 2),
                 ranges = IRanges(start = c(1, 2), end = c(1, 12)),
                 strand = rep("+", 2), names = rep(seqname[1],2))

grl <- GRangesList(grIn1) # <- make it GRangesList
names(grl) <- "tx1"

# Then we are ready to find the ORFs, they will be returned as GRangesList,
# grouped by which transcript they came from

test_ranges <- find_in_frame_ORFs(grl,seqs) # <- ORFs are now in test_ranges

```

# BSGenome example

```r
library(ORFik)
library(GenomicFeatures)

# Make BSGenome  example with fasta file and gtf file

# First the gtf file, we will extract the 5' leaders to find all upstream open reading frames.

# here we use example data from GenomicFeatures,
# usually you have your own txdb object, a gtf file etc.

samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite", package = "GenomicFeatures")
txdb <- loadDb(samplefile)
fiveUTRs = fiveUTRsByTranscript(txdb, use.names = TRUE) # <- extract only 5' leaders

# Now get the Fasta file, we of course choose the fasta file for hg19,
# that is version 19 of human genome

if(requireNamespace("BSgenome.Hsapiens.UCSC.hg19")){ # <- safe require for vignette

  tx_seqs <- extractTranscriptSeqs(Hsapiens, fiveUTRs) # <- Now get only fiveUTRs fastasequences
  test_ranges <- find_in_frame_ORFs(fiveUTRs,tx_seqs) # <- ORFs are now in test_ranges
  View(test_ranges)
}
```

You will see that test_ranges are grouped by transcript, the first group is from transcript 3.
There are 7 exons in this group, named 3_* where star is the orf number.
You will also see that the second orf named 3_2 is twice,
because that single orf is split on two exons.

# Cage-data for tissue specific 5' utrs

In the previous example we used the reference annotation of the 5' utrs
from the package GenomicFeatures.
This is a general reference, which will fail to catch tissue specific orfs.
Here is an example how to include cage-data, to set new Transcription start sites(TSS).

```r
library(ORFik)
library(GenomicFeatures)

# Make BSGenome  example with fasta file and gtf file

# First the gtf file, we will extract the 5' leaders and reassign them

# here we use example data from GenomicFeatures,
# usually you have your own txdb object, a gtf file etc.
  
samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite",
                package = "GenomicFeatures")
txdb <- loadDb(samplefile)
fiveUTRs = fiveUTRsByTranscript(txdb, use.names = TRUE) # <- extract only 5' leaders
cageData <- system.file("extdata", "cage_data_heart.bed.bgz",
                package = "ORFik") # location of the cage file
extension <- 1000 # how much upstream of the original tss can the new tss be ?

filterValue <- 2 # filter for max cage-peak, how many reads must the max peak have
# to be valid ?
cds <- cdsBy(txdb, by = "tx", use.names = TRUE) # if you want to find 
# upstream open reading frames (uORFs), that can go into the cds.
changedFiveUTRs <- reassignTSSbyCage(fiveUTRs, cageData, extension,
                                        filterValue, cds)
View(changedFiveUTRs)

```

You will now see that most of the transcription start sites have changed.
For different tissues, you will get different leaders.
From testing on Fantom5 cage-data, less than 1% of the start sites remain the same
on average after reassignment by cage.

# Finding tissue specific ORFs
Using the results from the last example, lets find the tissue specific uORFs from
the example heart cage-data.

```r
# From the previous two examples we had Hsapiens and changedFiveUTRs.
# we now assign to null to avoid vignette error.
changedFiveUTRs <- NULL # get this object from previous example
Hsapiens <- NULL # get this from the BSGenome example

tx_seqs <- extractTranscriptSeqs(Hsapiens, changedFiveUTRs) # <- Now get only fiveUTRs fastasequences

# run this:
#uORFs <- find_in_frame_ORFs(changedFiveUTRs,tx_seqs)
#View(uORFs)
```
The uORFs from this example will most likely contain some uORFs not in the original
example, also some will be lost. This is, because different tissues transcribes
different mRNAs.

# Syntax of Orfs in ORFik

The ORFs in ORFik are grouped in a GRangesList. Each group is the exons per orf.
The names of the GRangesList is the orf names, named as transcriptName_*, 
where transcriptName is the transcriptName i.g for ensembl: "ENST123101239".
The "_" is the seperator
The * is an integer, where the first orf found on transcript is "_1", second is "_2" etc.
The first orf will alway be the one most upstream for + strand, and least upstream 
for - strand
For a GRangesList called grl with 1 orf, names(grl) will return i.g. "ENST123101239_1"
Lets say the orf have 2 exons, if you unlist the grl,
then names(unlist(grl, use.names = FALSE)) will return
the transcript names, i.g. c("ENST123101239", "ENST123101239")
The unlisted orf also have a metacolumn called names, which contains the orf names,
i.g. c("ENST123101239_1", "ENST123101239_1") for this example.
The reason for this, is to keep redundant information for easy unlisting.

# Features from ORFs

ORFik contains a big feature set:
FLOSS (?floss), coverage (?coverage) , ORFscore (?ORFScore), entropy (?entropy) , 
translational effiency (?te), inside outside score (?insideOutsideScore),
distance between orfs and cds' (?distOrfToCds) etc.
All the features are justified with a reference to an article, i.g to see
justification for floss, see ?floss


Now lets do an example to get all features.

```r

# From the previous two examples we had Hsapiens and changedFiveUTRs.
# we now assign to null to avoid vignette error.
# additionally we need the GenomicAlignments package
if(requireNamespace("GenomicAlignments")){ # <- safe require for vignette
  changedFiveUTRs <- NULL # get this object from previous example
  Hsapiens <- NULL # get this from the BSGenome example, a FaFile.
  txdb <-NULL # the gtf file as TxDb
  uORFs <- NULL # the uorfs from previous example
  fiveUTRs <- NULL # the original non cage 5' utrs
  
  
  RFP <- import.bed(system.file("extdata", "ribo-seq-heart.bed.bgz",
                  package = "ORFik")) # can also be bam
  RNA <- import.bed(system.file("extdata", "rna-seq-heart.bed.bgz",
                  package = "ORFik")) # can also be bam
  # important thing  here! if RFP is not shifted for the p-site,
  # the scores will be wrong, see example of p-site shifting.
  # You can use ORFik to do this.
  riboStart <- 26 # minimum allowed fragment length on ribo-seq data.
  riboStop <- 34 # maximum allowed fragment length on ribo-seq data.
  extension <- 1000 # in the original cage-data example, we extended the
  # 5' utrs by a 1000 upstream, add this as information.
  
  # run this:
  #allFeatures <- allFeatures(grl = uORFs, orfFeatures = T, RFP, RNA,  Gtf = txdb,
  #  fiveUTRs = fiveUTRs, faFile = Hsapiens, riboStart = riboStart, 
  #   riboStop = riboStop, extension = extension)
  #View(allFeatures)
}
```

You will now get a data.table with one column per score, the columns are named after
the different scores, you can now go further with prediction, or making plots.

# Grouping ORFs

Sometimes you want a GRangesList of orfs grouped by transcript, or you might want 
each orf as groups in the GRangesList. To do this more easily you can use the function
ORFik:::groupGRangesBy.

```r
# the first part is equal to BSGenome example:

library(ORFik)
library(GenomicFeatures)

samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite", package = "GenomicFeatures")
txdb <- loadDb(samplefile)
fiveUTRs = fiveUTRsByTranscript(txdb, use.names = TRUE) # <- extract only 5' leaders

if(requireNamespace("BSgenome.Hsapiens.UCSC.hg19")){ # <- safe require for vignette

  tx_seqs <- extractTranscriptSeqs(Hsapiens, fiveUTRs) # <- Now get only fiveUTRs fastasequences
  test_ranges <- find_in_frame_ORFs(fiveUTRs,tx_seqs) # <- ORFs are now in test_ranges
  
  # test_ranges is now grouped by transcript, but we want them grouped by ORFs:
  # we use the orfs exon column called ($names) to group, it is made by ORFik.
  test_ranges <- groupGRangesBy(test_ranges, test_ranges$names)
  
  # the orfs are now grouped by orfs. If we want to go back to transcripts we do:
  
  test_ranges <- groupGRangesBy(test_ranges,
    unique(names(unlist(test_ranges, use.names = FALSE)))
}
```

# Aligning ribo-seq data centered on p-sites

Features like floss needs the ribo-seq data to be alligned for the p-site.
To do this, you can either do it manually by using plots from ORFik, to 
deduce the p-sites, or you can do it automaticly by using the package 
shoelaces(recomended). 

# Predicting functional orfs

If you have found a set of orfs, and you have used the function ORFik:::allFeatures
to find the features for those orfs, you can try to predict which orfs to might be coding.
To do this you can specify some cut-off for the features, so that most of the orfs get excluded.

An example could be:

For all orfs with ribo-seq fpkm < 1, filter out.
For all orfs with kozak score < 5, filter out.
For all orfs with ORFScore < 0.5, filter out.

The more strict you are, the less orfs you get. This smaller set can
be verified further with other ORF-tools, experiments or by proteomics.

Now an example on Upstream open reading frames(uorfs):

For all uorfs with ribo-seq fpkm bigger than coding sequences fpkm, keep them.

Even though the uorf is translated, it does not mean it regulates a gene-product.
Two ways to check this: 

One of the features is if the orf goes into the cds of its transcript, 
you can use it like this:
For all uorfs with distance between orf and cds > 0,
(which mean it does not goes into cds), filter out.
Now you will only have uorfs going into cds.

For all uorfs with distance to cds > 200, filter out.
Since the distance is so high, the ribosome most likely can 
reinitiate. But this also depends on the region between the uorf stop site, and
coding sequence start site.

Because of the small set of experimentally verified uorfs, it can be hard to know
what are good features for uorfs, ribo-seq features like the ones in this package, 
at least tries to predict what could be relevant.
